Awful.apk/src/main/java/com/ferg/awfulapp/ForumsIndexActivity.kt

```diff
@@ -0,0 +1,276 @@
+/**
+ * *****************************************************************************
+ * Copyright (c) 2011, Scott Ferguson
+ * All rights reserved.
+ *
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ * * Redistributions of source code must retain the above copyright
+ * notice, this list of conditions and the following disclaimer.
+ * * Redistributions in binary form must reproduce the above copyright
+ * notice, this list of conditions and the following disclaimer in the
+ * documentation and/or other materials provided with the distribution.
+ * * Neither the name of the software nor the
+ * names of its contributors may be used to endorse or promote products
+ * derived from this software without specific prior written permission.
+ *
+ *
+ * THIS SOFTWARE IS PROVIDED BY SCOTT FERGUSON ''AS IS'' AND ANY
+ * EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+ * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL SCOTT FERGUSON BE LIABLE FOR ANY
+ * DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
+ * ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ * *****************************************************************************
+ */
+
+package com.ferg.awfulapp
+
+import android.app.Activity
+import android.content.Intent
+import android.content.res.Configuration
+import android.net.Uri
+import android.os.Bundle
+import android.support.design.widget.Snackbar
+import android.support.v7.widget.Toolbar
+import android.view.KeyEvent
+import android.view.View
+
+import com.ferg.awfulapp.announcements.AnnouncementsManager
+import com.ferg.awfulapp.constants.Constants
+import com.ferg.awfulapp.dialog.Changelog
+import com.ferg.awfulapp.messages.PmManager
+import com.ferg.awfulapp.preferences.AwfulPreferences
+import com.ferg.awfulapp.preferences.Keys
+import com.ferg.awfulapp.sync.SyncManager
+
+import java.util.Locale
+
+import timber.log.Timber
+
+
+class ForumsIndexActivity :
+        AwfulActivity(),
+        PmManager.Listener,
+        AnnouncementsManager.AnnouncementListener {
+
+    private lateinit var forumsPager: ForumsPagerController
```

Comment:
lateinit doesn't work for the pager controller and the nav drawer, causes crashes when the app recreates (turn on Don't keep activities in dev options)

Basically the various fragments can get recreated before the activity hits onCreate, which is where these things get initialised. When they call the activity to update it about their state, you need to ensure the nav drawer and pager are ready - there's a race condition when you're restoring the app, and often they're not ready before the fragments

The original version does null checks on those two components, and basically ignores the update call if they're not ready. The forums pager will eventually initialise, it'll recover those fragments and add them to its viewpager, and then it'll inform the activity that the current page has updated through PagerCallbacks (another reason not to delete it!)

So yeah, unfortunately we'll have to keep these nullable and initialise them as null, and do null checks where they existed in the original java version. Most of the other calls will happen after onCreate though, so we can !! those

Reply:
Thanks for the note about the dev option. I always forget about that and I definitely need to be testing with it on.

On your recommendation, I mostly agree although now that I know there is a race condition there I'm concerned about other edge cases that could show up - lateinits are really prickly if they aren't safely initialized first. That said, you're way more familiar with the app architecture than me so I'll go with that, with the expectation that it could change in the future. I'm making an `activityInitialized`  field in `AwfulActivity` - I could put it in `ForumsIndexActivity` but it seems like it'll be useful for other potential race conditions like this so it might as well go in the base object, I can move it if you don't agree.